### PR TITLE
Use stdatomic.h atomic primitives

### LINKED ISSUE #345: 'OSMemoryBarrier' is deprecated in macOS 10.12
Building with latest XCode I got lots of these warnings:

```
warning: 'OSMemoryBarrier' is deprecated: first deprecated in macOS 10.12 - Use std::atomic_thread_fence() from <atomic> instead [-Wdeprecated-declarations]
warning: 'OSAtomicAdd32Barrier' is deprecated: first deprecated in macOS 10.12 - Use std::atomic_fetch_add() from <atomic> instead [-Wdeprecated-declarations]
warning: 'OSAtomicCompareAndSwap64Barrier' is deprecated: first deprecated in macOS 10.12 - Use std::atomic_compare_exchange_strong() from <atomic> instead [-Wdeprecated-declarations]
```

### PR DESCRIPTION
**Pull request by ebraminio**
Tries to fix #345, `hb_atomic_ptr_impl_cmpexch` part is not ok

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -71,7 +71,7 @@ endif ()
 
 ## Detect if we are running inside a distribution or regular repository folder
 set(IN_HB_DIST FALSE)
-if (EXISTS "${PROJECT_SOURCE_DIR}/src/hb-version.h")
+if (EXISTS "${PROJECT_SOURCE_DIR}/ChangeLog")
```
Comment (ebraminio) [inline CMakeLists.txt]: Better is-in-distribution detection

File: CMakeLists.txt [hunk 2]
```diff
@@ -310,12 +310,29 @@ endif ()
...
 
 ## Atomic ops availability detection
+file(WRITE "${PROJECT_BINARY_DIR}/try_compile_stdatomic.c"
+"
+  #include <stdatomic.h>
+  atomic_int value = ATOMIC_VAR_INIT(1);
+  int main() {
+  int loaded_value = atomic_load(&value);
+    return 0;
+  }
```
Comment (ebraminio) [inline CMakeLists.txt]: src: https://bugs.python.org/file36151/atomicv2.patch

File: configure.ac [hunk 3]
```diff
@@ -441,6 +441,24 @@ if $have_coretext; then
...
+
+AC_CACHE_CHECK([for C compiler stdatomic.h], hb_cv_have_stdatomic, [
+	hb_cv_have_stdatomic=false
+	AC_TRY_LINK([
+		#include <stdatomic.h>
+		atomic_int value = ATOMIC_VAR_INIT(1);
+		int main() {
+			int loaded_value = atomic_load(&value);
+			return 0;
+		}
```
Comment (ebraminio) [inline configure.ac]: src: https://bugs.python.org/file36151/atomicv2.patch

File: src/hb-atomic-private.hh [hunk 4]
```diff
@@ -46,6 +46,17 @@
...
+#elif HAVE_STDATOMIC
+
+#include <stdatomic.h>
+
+typedef atomic_int hb_atomic_int_impl_t;
+#define HB_ATOMIC_INT_IMPL_INIT(V) (V)
+#define hb_atomic_int_impl_add(AI, V)		atomic_fetch_add (&(AI), (V))
+
+#define hb_atomic_ptr_impl_get(P)		(atomic_thread_fence (memory_order_release), (void *) *(P))
+#define hb_atomic_ptr_impl_cmpexch(P,O,N)	(atomic_compare_exchange_strong ((atomic_int *) (O), (int *) (N), (size_t)(P)))
```
Comment (ebraminio) [inline src/hb-atomic-private.hh]: @behdad: This needs a closer look. Does it look OK?

File: src/hb-atomic-private.hh [hunk 5]
```diff
@@ -46,6 +46,17 @@
...
+#elif HAVE_STDATOMIC
+
+#include <stdatomic.h>
+
+typedef atomic_int hb_atomic_int_impl_t;
+#define HB_ATOMIC_INT_IMPL_INIT(V) (V)
+#define hb_atomic_int_impl_add(AI, V)		atomic_fetch_add (&(AI), (V))
+
+#define hb_atomic_ptr_impl_get(P)		(atomic_thread_fence (memory_order_release), (void *) *(P))
+#define hb_atomic_ptr_impl_cmpexch(P,O,N)	(atomic_compare_exchange_strong ((atomic_int *) (O), (int *) (N), (size_t) (P)))
```
Comment (ebraminio) [inline src/hb-atomic-private.hh]: @behdad: These need a closer look

**Comment by behdad** [comment]
Thanks Ebrahim.

**Comment by ebraminio** [comment]
Far from perfect, perhaps needs your help
* autotools doesn't seem is able to detect stdatomic.h correctly on Linux
* hb-atomic-private.hh logic is not correct

**Comment by ebraminio** [comment]
Facing [more and more issues](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60932) with the solution, closing it, at least for now.

**Comment by khaledhosny** [comment]
What about using the C++11 `<atmic>`?

**Comment by ebraminio** [comment]
With C++11 <atomic>`
```
#include <atomic>

typedef std::atomic_int hb_atomic_int_impl_t;
#define HB_ATOMIC_INT_IMPL_INIT(V) {V}
#define hb_atomic_int_impl_add(AI, V)		std::atomic_fetch_add (&(AI), (V))

#define hb_atomic_ptr_impl_get(P)		(std::atomic_thread_fence (std::memory_order_release), (void *) *(P))
#define hb_atomic_ptr_impl_cmpexch(P,O,N)	(std::atomic_compare_exchange_strong ((std::atomic<int> *) (P), (int *)(O), (N)))
```
Honestly, can not figure how to make `hb_atomic_ptr_impl_cmpexch` work :(

**Comment by behdad** [comment]
> Honestly, can not figure how to make hb_atomic_ptr_impl_cmpexch work :(

`atomic<void *>` doesn't work?

**Comment by behdad** [comment]
> Honestly, can not figure how to make hb_atomic_ptr_impl_cmpexch work :(

`atomic<void *>` doesn't work?

> #define hb_atomic_ptr_impl_get(P) (std::atomic_thread_fence (std::memory_order_release), (void *) *(P))

I think this should be `memory_order_acquire`, or `_consume`, NOT `_release`.

**Comment by ebraminio** [comment]
```C++
#include <atomic>

typedef std::atomic<void *> hb_atomic_int_impl_t;
#define HB_ATOMIC_INT_IMPL_INIT(V) {V}
#define hb_atomic_int_impl_add(AI, V)		std::atomic_fetch_add (&(AI), (V))

#define hb_atomic_ptr_impl_get(P)		(std::atomic_thread_fence (std::memory_order_acquire), (void *) *(P))
#define hb_atomic_ptr_impl_cmpexch(P,O,N)	(std::atomic_compare_exchange_strong ((std::atomic<void *> *) (P), (void *) (O), (void *) (N)))
```

Not there, yet :(

```
./hb-atomic-private.hh:191:34: error: cannot initialize return object of type
      'int' with an rvalue of type 'void *'
  ...hb_atomic_int_impl_add (const_cast<hb_atomic_int_impl_t &> (v),  1); }
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by behdad** [comment]
Ebrahim,

You are missing one detail.  The following, should be written using `atomic<int>`:
```

typedef std::atomic<int> hb_atomic_int_impl_t;
#define HB_ATOMIC_INT_IMPL_INIT(V) {V}
#define hb_atomic_int_impl_add(AI, V)		std::atomic_fetch_add (&(AI), (V))
```
while the other two using `atomic<void *>`:
```
#define hb_atomic_ptr_impl_get(P)		(std::atomic_thread_fence (std::memory_order_acquire), (void *) *(P))
#define hb_atomic_ptr_impl_cmpexch(P,O,N)	(std::atomic_compare_exchange_strong ((std::atomic<void *> *) (P), (void *) (O), (void *) (N)))
```

Eventually I like to just use atomic.  But that will be a long time in the future...

**Comment by ebraminio** [comment]
Still not working as third parameter of atomic_compare_exchange_strong as it takes in a value rather on expected than pointer, unlike the second parameter and so it wouldn't accept a (void *) AFAIK. Perhaps you should get your hands dirty on this :)

**Comment by ebraminio** [comment]
Also please note that I used {V} instead (V) on C++ version and unlike rest of the atomics, perhaps should or shouldn't be done, I am not sure.

**Comment by behdad** [comment]
> Perhaps you should get your hands dirty on this :)

Yeah was going to, but there's more important stuff and I feel bad not working on them :(.  Let this be here and I'll eventually figure it out.  Thanks!

**Comment by behdad** [comment]
> Also please note that I used {V} instead (V) on C++ version

Use `ATOMIC_VAR_INIT` instead?

**Pull request closed without merging**